### DISCUSSION TITLE
OTP code autofill

### DISCUSSION THREAD

**Discussion by raresn** [airsdk/Adobe-Runtime-Support#3060] [General]
Hey @ajwfrost . Wondering if we can get this functionality in AIR:
https://stackoverflow.com/questions/39631168/automatic-otp-verification-in-ios
People get OTP's all the time and would be nice to have a modern approach to this in AIR.
ps not only ios, but Android as well.

**Comment by ajwfrost** [comment]
If I read that right, it sounds like we could just set a property on a StageText implementation and it would then just magically work...
Will have to get one of the guys to try this!

**Comment by raresn** [comment]
@ajwfrost thx for the beyond-fast reply!

Super cool!
Not sure about android tho` - could this be a similar thing?
Is StageText the only thing we care about in a Starling/Feather enviroment? (just asking)

**Reply by ajwfrost** [reply]
Not quite sure how Feathers would handle editable text fields, hopefully it would create a StageText instance over the top of the built-in/AIR element so that you get device-specific capabilities.

On the iOS side of things .. if you have the text field focused, and an SMS arrives, you get a "From messages: 012345" alert at the bottom of the screen, click in it and the code gets copied into the text field.

For Android, there seems to be some autofill hints:
https://developer.android.com/reference/android/view/View#setAutofillHints(java.lang.String[])
and as long as we set this up, and the user has it enabled:
https://stackoverflow.com/a/71322470

Any this also then works if we set up the StageText field appropriately.

Will need a new API for this I guess, we can squeeze it in to the (much-delayed!) AIR 51 as it's fairly straightforward.
On Android though, this is handled via "autofill" whereas on iOS it's a "content type".. and it seems there's a bit of an overlap between this and the soft keyboard type. But if we change it on Android to hint for an email address autofill, it doesn't change the keyboard at all, but would presumably try to auto-complete from previously known email addresses. On iOS it doesn't seem to have this auto-fill option but does have a type which would change the keyboard and maybe other related input mechanisms, it says -> so perhaps it also does this where possible.

Maybe we can just give the options and see what happens in each case :-)
e.g. `stageText.contentType = StageTextContentType.SMS_OTP;`
and `stageText.contentType = StageTextContentType.EMAIL_ADDRESS;`

**Reply by raresn** [reply]
@ajwfrost sounds lovely - let me know if i can help in any way.

**Reply by raresn** [reply]
@ajwfrost 'stageText.contentType = StageTextContentType.SMS_OTP;' does not produce any results when we get an sms called "your code is 1234"
We tested this on Android. Do we need to add anything to the manifest?
How did you test? Any steps to replicate a working behaviour?

**Reply by raresn** [reply]
@ajwfrost i know you have plenty on your plate - just leaving this here as a reminder.

**Reply by ajwfrost** [reply]
Thanks - sorry, had forgotten to reply! This feature was a "work in progress" -> there's still some glue work we need to do under the hood. I think it's half-done though, will check later and hopefully it will be in 51.0.0.3 shortly....
thanks

**Reply by ajwfrost** [reply]
@raresn can you please retry with the latest version: https://airsdk.harman.com/previews
thanks

**Reply by raresn** [reply]
@ajwfrost nothing happens. The text field is there, is in focus, and I got an SMS with "your code is 1234"
Do i need to do anything else?

**Reply by raresn** [reply]
how i implemented:
I have a class:
```
public class OTPStageTextTextEditor extends StageTextTextEditor
	{
		public function OTPStageTextTextEditor()
		{
			super();

		}
		public function set contentType(value: String)
		{
			if(stageText is StageText)
			{
				var st:StageText = stageText as StageText;
				st.contentType = value
			}

		}
		public function get contentType():String
		{
			if(stageText is StageText)
			{
				var st:StageText = stageText as StageText;
				return st.contentType;
			}
			return null;

		}
	}
```

**And then i have this:**
```
	input = new TextInput();
			input.textEditorFactory = function():ITextEditor {
				var editor:OTPStageTextTextEditor = new OTPStageTextTextEditor();
				editor.contentType = StageTextContentType.OTP_SMS;
				return editor;
			};
```
and i am using this input on stage.

**Reply by ajwfrost** [reply]
That looks okay to me.. I had found that the StageText needed to be empty, before it prompted me to paste in the value.

A couple of things to check perhaps:
1) is there any way that you can check the value has been set up properly, i.e. reading it back after that factory method has been called? not sure what might happen if 'stageText' is null, for example .. or delayed-initialisation things, could cause issues?
2) If this is on Android, have you also set up the account settings for it? e.g. "Settings -> Google -> Autofill -> SMS Verification codes -> Switch to enabled" or something similar in newer versions perhaps?

We'd checked it on both Android and iOS.. wondering if there's a way for us to add some tracing/debugging if that's not working for you, just to see where it's going wrong..

thanks

**Reply by raresn** [reply]
@ajwfrost  1  was the problem. Fixed it.
So, i started the app, 
- when selected the input field i got "Autofill code from messages". 
- it said waiting for messages.
- i got the message, the field got deselected (which is wrong)
- I select the field back, and i have the option to paste the code on the keyboard

This is nice, but not exactly how we imagined this working.
In other apps, this gets autocompleted, without any other interaction from users. Do you know what i`m talking about? Maybe i can send a video from another app?

Having to press autofill code, then select the field again, then paste.. it's a bit too much for the purpose of this.
(Didn't have an iphone yet to test this on)

**Reply by raresn** [reply]
https://developers.google.com/identity/sms-retriever/overview

**Reply by ajwfrost** [reply]
Hi
On my Samsung phone, what happens is:
1) I have the app open
2) I receive a text message (notification comes across the top of the screen but I can ignore it)
3) Click in the StageText field and the soft keyboard pops up
4) At the top of the keyboard, where the suggestions/autocomplete usually sits, is the OTP that I can click on to enter it into the text field

So, that's the OTP SMS hint mechanism that we'd looked at; the SMS-Retriever functionality that you link to above would be something else entirely - and one that needs the app to coordinate things with a server i.e. it's not just a case of any SMS with the word "code" plus a 5-6 digit number... it would need to include the application hash in the message from the server, and it's very Android-specific. If that's what you're after then this is something we could put together as an ANE. It wouldn't involve any StageText entry though, it's all about phone number verification behind the scenes...?

thanks

**Reply by raresn** [reply]
@ajwfrost i understand what you mean, and yes, I was thinking of two things at once. So this OTP works for it's original purpose, but the steps change from one phone to another (i have a samsung s23 here). Hope i`ll be able to test on iOS today as well. 
Of course an ANE that would do the SMS-Retriever for iOS & Android would be absolutely superb. Is that something that can really happen from Harman? or how?
Thanks!
